src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs

```diff
@@ -110,6 +145,44 @@ public override void WriteTo(Utf8JsonWriter writer, JsonSerializerOptions? optio
             }
         }
 
+        internal override bool DeepEquals(JsonNode? node)
+        {
+            if (node is null || node is JsonArray)
+            {
+                return false;
+            }
+
+            if (node is JsonValue value)
+            {
+                // JsonValueTrimmable/NonTrimmable can hold the object type so calling this method to continue the deep comparision.
+                return value.DeepEquals(this);
+            }
+
+            JsonObject jsonObject = node.AsObject();
```

Comment:
IIRC `AsObject` just checks if `node` is `JsonObject` and otherwise throws. I think all that logic could be condensed into a switch statement like so:
```C#
switch (node)
{
     case null or JsonArray: 
          return false;
     case JsonValue value:
          /* */
      case JsonObject:
        /* */
     default:
          Debug.Fail("Impossible case");
          return false;
}
```

Reply:
Same applies to the `DeepEquals` implementation in `JsonArray` obviously.